drivers/usbhost/usbhost_cdcmbim.c

```diff
@@ -0,0 +1,2562 @@
+
+  if (priv->bulkinbytes < (int16_t)(sizeof(struct usb_cdc_ncm_nth16_s) +
+                                    sizeof(struct usb_cdc_ncm_ndp16_s)))
+    {
+      goto out;
+    }
+
+  /* Parse the NTB header */
+
+  struct usb_cdc_ncm_nth16_s *nth =
+    (struct usb_cdc_ncm_nth16_s *)priv->rxnetbuf;
+
+  if (usbhost_getle32(nth->signature) != USB_CDC_NCM_NTH16_SIGNATURE)
+    {
+      uerr("Invalid NTH signature\n");
+      goto out;
+    }
+
+  uint16_t block_len = usbhost_getle16(nth->block_length);
+
+  if (block_len > priv->bulkinbytes)
+    {
+      uerr("Block length larger than rx buffer\n");
+      goto out;
+    }
+
+  ndpoffset = usbhost_getle16(nth->ndp_index);
+  if (ndpoffset > priv->bulkinbytes)
+    {
+      uerr("NDP offset too far %u > %u\n", ndpoffset, priv->bulkinbytes);
+      goto out;
+    }
+
+  /* Parse each NDP */
+
+  do
+    {
+      struct usb_cdc_ncm_ndp16_s *ndp
+        = (struct usb_cdc_ncm_ndp16_s *)(priv->rxnetbuf + ndpoffset);
+
+      ndpoffset = usbhost_getle16(ndp->next_ndp_index);
+
+      /* Parse each DPE */
+
+      struct usb_cdc_ncm_dpe16_s *dpe;
+
+      for (dpe = ndp->dpe16; usbhost_getle16(dpe->index); dpe++)
+        {
+          dgram_off = usbhost_getle16(dpe->index);
+          dgram_len = usbhost_getle16(dpe->length);
+
+          if (dgram_off + dgram_len <= priv->bulkinbytes)
+            {
+              cdcmbim_receive(priv, priv->rxnetbuf + dgram_off, dgram_len);
+            }
+        }
+    }
+  while (ndpoffset);
+
+out:
+    DRVR_ASYNCH(hport->drvr, priv->bulkin,
+                (uint8_t *)priv->rxnetbuf, CDCMBIM_NETBUF_SIZE,
+                usbhost_bulkin_callback, priv);
+    usbhost_givesem(&priv->exclsem);
+}
+
+/****************************************************************************
+ * Name: usbhost_rxdata_work
+ *
+ * Description:
+ *   Read an available comm message into the RX buffer
+ *
+ * Input Parameters:
+ *   arg - Driver private data
+ *
+ * Returned Value:
+ *   None
+ *
+ * Assumptions:
+ *   None
+ *
+ ****************************************************************************/
+
+static void usbhost_rxdata_work(FAR void *arg)
+{
+  FAR struct usbhost_cdcmbim_s *priv;
+  int ret;
+
+  priv = (FAR struct usbhost_cdcmbim_s *)arg;
+  DEBUGASSERT(priv);
+
+  /* Are we still connected? */
+
+  if (priv->disconnected)
+    {
+      return;
+    }
+
+  usbhost_takesem(&priv->exclsem);
+
+  ret = usbhost_ctrl_cmd(priv,
+                         USB_REQ_DIR_IN | USB_REQ_TYPE_CLASS |
+                         USB_REQ_RECIPIENT_INTERFACE,
+                         USB_CDC_GET_ENCAPSULATED_RESPONSE,
+                         0, priv->ctrlif, priv->comm_rxbuf,
+                         priv->maxctrlsize);
+  if (ret)
+    {
+      uerr("Failed to read message: %d\n", ret);
+      goto errout;
+    }
+
+  /* Would be nice if CTRLIN would return how many bytes it read... */
+
+  struct mbim_header_s
+  {
+    uint8_t type[4];
+    uint8_t len[4];
+  };
+
```

Comment:
This is a C89 violation and must be fixed.  Move this structure to the top of the function.